### DISCUSSION TITLE
Certain wagmi hooks not working for WalletConnect V2

### DISCUSSION THREAD

**Discussion by crypblizz8** [WalletConnect/walletconnect-monorepo#2534] [Developer Support]
This could be a wagmi issue but posting it here first.
- posting on behalf of Venus Protocol
- Version 0.12 of wagmi
- Trying to use `getBalance` and `useSigner`
- V1 works perfectly with hooks such as useSigner.

Process
- In our dApp, we grab the signer using the `useSigner` hook from wagmi, then we create a Contract instance (Contract being the class from ethers.js) using the address and abi of the contract and passing it the signer mentioned earlier.

Minimal Example Repo
- https://github.com/therealemjy/wallet-connect-v2-signer-issue-example

How to solve this:
- Can we investigate if this is an issue from our side or just wagmi related?

![telegram-cloud-photo-size-1-5152459750485634033-y](https://github.com/WalletConnect/walletconnect-monorepo/assets/45455218/5fbbb072-604e-4385-b6ea-98491ed7b3d1)

**Comment by crypblizz8** [comment]
CC @0xAsimetriq  / @boidushya

**Reply by xzilja** [reply]
This usually means that `eth_call` method was not approved by wallet in optional namespaces. Also, ideally they should update to latest wagmi.

**Reply by boidushya** [reply]
Updating to latest version of wagmi w/ viem fixes this for me.
Spec Details:

wagmi v1.1.0
viem v0.3.49

chains: [mainnet] // even though I tried this for ETH Mainnet, this should work with BSC Testnet in a same way

<img width="1510" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/26746725/900e53eb-0369-4ec2-ac04-1dfe7c5c7b6f">

**Comment by boidushya** [comment]
As a temporary fix, you can fetch balance using something like this with wagmi v0.12.x and ethers v5

```typescript
  const { address } = useAccount();
  const { data: balance } = useBalance({
    address,
  });
  const fetchBalance = () => {
    console.log("Balance", balance);
  };
```

Although, to quote @0xAsimetriq :
> This is again, optionalNamespaces, basically when you see that method error, it means wallet didn't specify they support such method in their session response.

**Reply by boidushya** [reply]
Adding a screenshot to show what the output looks like

<img width="1509" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/26746725/9e7ab36b-2d39-4709-a06c-32e4324ae40d">

**Comment by therealemjy** [comment]
@0xAsimetriq @boidushya 
I work for Venus, thanks for looking at this issue.

Unfortunately upgrading to the latest version of Wagmi isn't really possible now, as switching from `ethers` v5 to `viem` would require us to refactor a lot of core logic in our dApp.

Regarding the optional namespaces not being specific by the wallet: is that something you guys expect wallets to adapt to in the near future?

@boidushya Thanks for providing a workaround to fetch the user balance. I tried to generate a minimal example repo, but the real issue for us right now is making contract calls using a signer. Do you know a workaround for this?
To give you more context, this is typically how we get the signer from Wagmi and call contracts with it:
```typescript
import { useSigner } from "wagmi";
import { Contract } from "ethers";
import xvsAbi from "./abis/xvs.json";

const useGetAllowance = async ({
  accountAddress,
  spenderAddress,
}: {
  accountAddress: string;
  spenderAddress: string;
}) => {
  const { data: signer } = useSigner();
  const xvsContract = new Contract(
    "0xcF6BB5389c92Bdda8a3747Ddb454cB7a64626C63",
    xvsAbi,
    signer || undefined,
  );

  return xvsContract.allowance(accountAddress, spenderAddress);
};
```

**Reply by boidushya** [reply]
About optionalNamesapces, Wallets are implementing and QA'ing it at the moment, it should be done in the near future!

**Reply by therealemjy** [reply]
Ok, great! Thank you.
Then I think for now I'll wait for wallets to implement this.

**Reply by boidushya** [reply]
Also, the inability to make contract calls using a signer right now is related to the optionalNamespaces issue as well

**Reply by therealemjy** [reply]
Got you, thanks.

**Reply by finessevanes** [reply]
> Ok, great! Thank you. Then I think for now I'll wait for wallets to implement this.

What wallet/s were you testing with?

**Reply by therealemjy** [reply]
> > Ok, great! Thank you. Then I think for now I'll wait for wallets to implement this.
> 
> 
> 
> What wallet/s were you testing with? 

AlphaWallet and Venly.

**Comment by glitch-txs** [comment]
hi @0xAsimetriq! Would this be possible?

```ts
const options = { // customOptions for WalletConnect connector }

const { publicClient } = configureChains(chains, [w3mProvider({ projectId, options })])
```

I was thinking about hardcoding w3mProvider but [the wagmi connector doesn't let you set custom required events/methods](https://github.com/wagmi-dev/references/blob/827df8a2f8788711cad291c1e9b8c858c2ccc52f/packages/connectors/src/walletConnect.ts#L287). Maybe they could add an option for this? or a spread operator at the end of the options?

I also found this issue with Rainbow that's related https://github.com/rainbow-me/rainbow/issues/4872

**Reply by xzilja** [reply]
I think its a good idea to allow users to define custom optional namespaces 👍 (by default wagmi does add all methods and events however, so it shouldn't be an issue, wallet will return only what they support).

Regarding rainbow question, could you expand on what the issue is? Wagmi proposes 1 (currently selected) chain as required and all other dapp defined chains as optional to rainbow. They support optional namespaces now and return all the ones you can use as expected.
Below is video connecting with rainbow and using all their supported chains and it indicates unsupported ones (does this help, or is issue something else?):


https://github.com/WalletConnect/walletconnect-monorepo/assets/3154053/82b1b2a9-7a21-41e4-8175-8e1d08ac8f4d

**Reply by glitch-txs** [reply]
Mmm I see. I have deleted Rainbow and reinstalled and I get the same issue on Android. Also could you try scanning from PC?

https://github.com/WalletConnect/walletconnect-monorepo/assets/66949816/6d1168ea-5747-4945-8d33-bb745513e8a8

**Reply by xzilja** [reply]
Ah, that explains it. I don't think they published optional namespace changes to android yet. Thus it only connects to one required chain. I recently ran into issue with some optional methods there as well.

**Comment by crypblizz8** [comment]
This was solved in the latest wagmi upgrade to `0.12.16`
